### ISSUE TITLE
False positive on S1172 when using async lambdas an local functions

### ISSUE THREAD

**Issue by jairbubbles** [SonarSource/sonar-dotnet#4199]
### Description

Function parameter is marked as unused while it's used through a lambda and a local function.

### Repro steps

Here is a simple case:

```csharp
private Task DoSomethingAsync(string text) // 'text' is reported as unused 
{
    return Task.Run(async () => await UseAsync());

    async Task UseAsync()
    {
        await Task.Delay(100);
        Console.WriteLine(text); // while it's used there
    }
}
```

### Known workarounds

We can pass the parameter explicitly to the local function:

```csharp
private Task DoSomethingAsync(string text)
{
    return Task.Run(async () => await UseAsync(text));

    async Task UseAsync(string text1)
    {
        await Task.Delay(100);

        Console.WriteLine(text1);
    }
}
```

The problem also disappears if we remove the lambda:

```csharp
private async Task DoSomethingAsync(string text)
{
    await UseAsync();

    async Task UseAsync()
    {
        await Task.Delay(100);
        Console.WriteLine(text);
    }
}
```

### Related information

* C#/VB.NET Plugins version 8.20.0.28934
* Visual Studio version 16.9.2
* MSBuild / dotnet version 5.0
* Operating System Windows 10

**Comment by costin-zaharia-sonarsource** [comment]
Hi @jairbubbles, thanks for the feedback. I can confirm the false positive.

**Comment by AraHaan** [comment]
Same also happens with constructors falsely getting flagged as unused when using the new short style ``new()`` on static variables of a class/structure.

Both of which can repro easily.

**Comment by pavel-mikula-sonarsource** [comment]
Hi @AraHaan ,

Can you please create a new issue with a small code snippet for the C# 9 new() case? It will have different root cause than this.

**Comment by AraHaan** [comment]
> Hi @AraHaan ,
> 
> Can you please create a new issue with a small code snippet for the C# 9 new() case? It will have different root cause than this.

I sure can.
